File: healthhub-backend/apps/scheduling/models/appointment.py
from django.conf import settings
from django.db import models

from apps.accounts.models import UserRole


class AppointmentStatus(models.TextChoices):
    """The lifecycle states of a scheduled appointment."""

    SCHEDULED = "scheduled", "Scheduled"
    CHECKED_IN = "checked_in", "Checked in"
    COMPLETED = "completed", "Completed"
    CANCELLED = "cancelled", "Cancelled"
    NO_SHOW = "no_show", "No show"


class Appointment(models.Model):
    """A patient's booked visit to a facility for specimen collection.

    Appointments are the patient-facing scheduling primitive. Test orders may
    reference the appointment at which their specimen is collected.
    """

    patient = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.CASCADE,
        related_name="appointments",
        limit_choices_to={"role": UserRole.PATIENT},
    )
    facility = models.ForeignKey(
        "directory.Facility",
        on_delete=models.PROTECT,
        related_name="appointments",
    )
    referring_doctor = models.ForeignKey(
        settings.AUTH_USER_MODEL,
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
        related_name="referred_appointments",
        limit_choices_to={"role": UserRole.DOCTOR},
    )
    scheduled_at = models.DateTimeField()
    status = models.CharField(
        max_length=20,
        choices=AppointmentStatus.choices,
        default=AppointmentStatus.SCHEDULED,
    )
    reason = models.CharField(max_length=255, blank=True)
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        ordering = ["-scheduled_at"]

    def __str__(self):
        """Return the patient, facility, and time for admin/reporting displays."""
        return (
            f"{self.patient.get_full_name()} @ {self.facility.name} "
            f"on {self.scheduled_at:%Y-%m-%d %H:%M}"
        )

File: healthhub-frontend/src/types/appointment.ts
/** Types for the scheduling (appointments) domain. */

import type { Facility } from "@/src/types/directory";

export type AppointmentStatus =
  | "scheduled"
  | "checked_in"
  | "completed"
  | "cancelled"
  | "no_show";

/** An appointment as returned by the API (read shape). */
export interface Appointment {
  id: number;
  patient: number;
  patient_name: string;
  facility: Facility;
  referring_doctor: number | null;
  referring_doctor_name: string | null;
  scheduled_at: string;
  status: AppointmentStatus;
  status_display: string;
  reason: string;
  created_at: string;
}

/** Payload for booking an appointment. `patient` is only sent by staff. */
export interface BookAppointmentInput {
  facility: number;
  scheduled_at: string;
  reason?: string;
  patient?: number;
  referring_doctor?: number;
}
